File: iosApp/iosApp/features/search_bar/presentation/SearchViewModel.swift
//
//  SearchViewmodel.swift
//  iosApp
//
//  Created by Dimitrios Katoudis on 3/11/25.
//

import Foundation
import ComposeApp

@MainActor
class SearchViewModel: ObservableObject {
    @Published var state: SearchState
    
    private let viewStore: SearchViewStore
    private let scope: Kotlinx_coroutines_coreCoroutineScope
    private var stateObservationTask: Task<Void, Never>?
    
    init() {
        self.scope = ScopeHelperKt.createMainScope()

        self.viewStore = KoinHelper.shared.iosHelper.getSearchViewStore(scope: scope)

        self.state = viewStore.searchState.value

        observeState()
    }
    
    private func observeState() {
        stateObservationTask = Task { @MainActor [weak self] in
            guard let stateFlow = self?.viewStore.searchState else { return }
            do {
                for try await newState in stateFlow {
                    guard let self else { return }
                    self.state = newState
                }
            } catch {
                print("Error observing state: \(error)")
            }
        }
    }
    
    func onEvent(_ event: SearchEvent) {
        viewStore.onEvent(event: event)
    }
    
    deinit {
        ScopeHelperKt.cancelScope(scope: scope)
        stateObservationTask?.cancel()
    }
    
}

File: web/searchCommon/src/jsMain/kotlin/org/example/fakeshop_clients/features/search/presentation/SearchViewModel.kt
package org.example.fakeshop_clients.features.search.presentation

import kotlinx.coroutines.MainScope
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.StateFlow

class SearchViewModel(
    private val store: SearchViewStore
) {
    val uiState: StateFlow<SearchState> = store.searchState

    private val scope = MainScope()

    fun onEvent(event: SearchEvent) {
        store.onEvent(event)
    }

    fun cleanup() {
        scope.cancel()
    }
}
